### PR TITLE
swarm/pss: Protocol RemovePeer

### PR DESCRIPTION
**Pull request by nolash**
https://github.com/ethersphere/go-ethereum/issues/529

Add method to "disconnect" peer,

### CODE REVIEW

File: swarm/pss/protocol.go [hunk 1]
```diff
@@ -233,6 +243,21 @@ func (self *Protocol) AddPeer(p *p2p.Peer, topic Topic, asymmetric bool, key str
 	return rw, nil
 }
 
+func (self *Protocol) RemovePeer(asymmetric bool, key string) {
```
Comment (acud) [inline swarm/pss/protocol.go:246]: if the test is the only consumer of this method - does it have to reside within the production code?

**Review by acud — APPROVED** [review]

File: swarm/pss/protocol.go (hunk 1 above)
Comment (nolash) [reply swarm/pss/protocol.go:246]: It won't be the only consumer. It will be used for connection control wrapper code in implementations.

**Review by nonsense — APPROVED** [review]

File: swarm/pss/client/client_test.go [hunk 2]
```diff
@@ -160,6 +161,15 @@ func TestClientHandshake(t *testing.T) {
 		log.Warn("ok", "idx", i, "got", got)
 		time.Sleep(time.Second)
 	}
+
+	rw := lpsc.peerPool[pss.PingTopic][rpubkey]
+	lpsc.RemovePssPeer(rpubkey, pss.PingProtocol)
+	if err := rw.WriteMsg(p2p.Msg{
+		Size:    3,
+		Payload: bytes.NewReader([]byte("foo")),
+	}); err == nil {
```
Comment (janos) [inline swarm/pss/client/client_test.go:170]: Just a suggestion to check for the explicit `connection close` error. This error could even have its own variable.

File: swarm/pss/protocol.go [hunk 3]
```diff
@@ -233,6 +243,21 @@ func (self *Protocol) AddPeer(p *p2p.Peer, topic Topic, asymmetric bool, key str
...
+	log.Debug("closing pss peer", "asym", asymmetric, "key", key)
+	self.RWPoolMu.Lock()
+	defer self.RWPoolMu.Unlock()
+	if asymmetric {
```
Comment (janos) [inline swarm/pss/protocol.go:250]: If asymmetric is of the wrong value, this function will panic. I am suggesting to ether check if pubKeyPool or symKeyPool rw should be closed, or to remove the asymmetric argument and self discover which rw to close.

File: swarm/pss/protocol_test.go [hunk 4]
```diff
@@ -130,5 +131,12 @@ func testProtocol(t *testing.T) {
 	case cerr := <-lctx.Done():
 		t.Fatalf("test message timed out: %v", cerr)
 	}
-
+	rw := pssprotocols[lnodeinfo.ID].protocol.pubKeyRWPool[rpubkey]
+	pssprotocols[lnodeinfo.ID].protocol.RemovePeer(true, rpubkey)
+	if err := rw.WriteMsg(p2p.Msg{
+		Size:    3,
+		Payload: bytes.NewReader([]byte("foo")),
+	}); err == nil {
```
Comment (janos) [inline swarm/pss/protocol_test.go:139]: Checking for the right error is nice to have, also with test to close peer with symmetric key.

Comment (zelig) [inline swarm/pss/protocol_test.go:139]: test the actual error you expect, not just that its not nil

**Pull request merged**